NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,464 @@
+#include "pch.h"
+#include <filesystem>
+#include <sstream>
+#include <fstream>
+#include "squirrel/squirrel.h"
+#include "util/utils.h"
+#include "mods/modmanager.h"
+#include "modsavefiles.h"
+#include "rapidjson/document.h"
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+#include "config/profile.h"
+#include "core/tier0.h"
+#include "rapidjson/error/en.h"
+
+SaveFileManager* g_saveFileManager;
+int MAX_FOLDER_SIZE = 52428800; // 50MB (50 * 1024 * 1024)
+fs::path savePath;
+
+uintmax_t GetSizeOfFolderContentsMinusFile(fs::path dir, std::string file)
+{
+	uintmax_t result = 0;
+	for (const auto& entry : fs::directory_iterator(dir))
+	{
+		if (entry.path().filename() == file)
+			continue;
+		result += fs::file_size(entry.path());
+	}
+	return result;
+}
+
+template <ScriptContext context> void SaveFileManager::SaveFileAsync(fs::path file, std::string contents)
+{
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread writeThread(
+		[mutex, file, contents]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				fs::path dir = file.parent_path();
+				// this actually allows mods to go over the limit, but not by much
+				// the limit is to prevent mods from taking gigabytes of space,
+				// this ain't a cloud service.
+				if (GetSizeOfFolderContentsMinusFile(dir, file.filename().string()) + contents.length() > MAX_FOLDER_SIZE)
+				{
+					// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
+					spdlog::error(fmt::format("Mod spamming save requests? Folder limit bypassed despite previous checks. Not saving."));
+					mutex.get().unlock();
+					return;
+				}
+
+				std::ofstream fileStr(file);
```

Comment:
I don't know where to put this, but file name encoding seems to fail with non-ASCII characters, as you can see in this screenshot: 

![image](https://user-images.githubusercontent.com/11993538/223225530-d1a5ab23-847f-44c1-9436-66998f4088d6.png)
Related squirrel code:
```javascript
NSSaveFile( "héllo.json", "héy" )
```

Note that loading the file (*i.e. `NSLoadFile( "héllo.json", [...])`) works correctly with non-ASCII characters.

Reply:
I dunno how to fix this

If anyone knows you're welcome to unresolve this

Reply:
Looks like a wrong encoding issue, but I don't think you can exactly change the encoding for the path?

Reply:
Why did you resolve this if the issue isn't fixed?
Maybe there are some streams that handle non-ASCII file names.

Reply:
Because I've no plans to fix it.

The file name may be wrong, but it decodes correctly, so I see no priority in fixing it.

Additionally, any research I have done on the issue was fruitless, so I don't see any point in trying to fix this at all ¯\_(ツ)_/¯

Reply:
I could make file names ASCII only to resolve this as well, but there's no real reason to do that outside of this bug.